// @gorhom/paper-onboarding/lib/commonjs/components/indicator/Indicator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNativeReanimated = _interopRequireDefault(require("react-native-reanimated"));

var _reactNativeSvg = require("react-native-svg");

var _lodash = _interopRequireDefault(require("lodash.isequal"));

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const {
  interpolate,
  Extrapolate
} = _reactNativeReanimated.default;

const AnimatedCircle = _reactNativeReanimated.default.createAnimatedComponent(_reactNativeSvg.Circle);

const BORDER_WIDTH = 2;

const IndicatorComponent = ({
  index,
  indicatorSize,
  indicatorBackgroundColor,
  indicatorBorderColor,
  animatedIndex,
  item
}) => {
  const radius = (0, _react.useMemo)(() => (indicatorSize - 2) / 2, [indicatorSize]); //#region animation

  const animatedRadius = interpolate(animatedIndex, {
    inputRange: [index - 1, index, index + 1],
    outputRange: [radius * 0.33, radius, radius * 0.33],
    extrapolate: Extrapolate.CLAMP
  });
  const animatedIconScale = interpolate(animatedIndex, {
    inputRange: [index - 1, index, index + 1],
    outputRange: [1 * 0.33, 1, 1 * 0.33],
    extrapolate: Extrapolate.CLAMP
  });
  const animatedIconOpacity = interpolate(animatedIndex, {
    inputRange: [index - 0.25, index, index + 0.25],
    outputRange: [0, 1, 0],
    extrapolate: Extrapolate.CLAMP
  });
  const animatedCircleFillOpacity = interpolate(animatedIndex, {
    inputRange: [index - 1, index],
    outputRange: [0, 1],
    extrapolate: Extrapolate.CLAMP
  }); //#endregion
  //#region styles

  const containerStyle = (0, _react.useMemo)(() => _objectSpread(_objectSpread({}, _styles.styles.container), {
    width: indicatorSize,
    height: indicatorSize
  }), [indicatorSize]);
  const iconStyle = (0, _react.useMemo)(() => _objectSpread(_objectSpread({}, _styles.styles.iconContainer), {
    left: BORDER_WIDTH * 2,
    right: BORDER_WIDTH * 2,
    top: BORDER_WIDTH * 2,
    bottom: BORDER_WIDTH * 2,
    borderRadius: indicatorSize,
    opacity: animatedIconOpacity,
    transform: [{
      scale: animatedIconScale
    }]
  }), // eslint-disable-next-line react-hooks/exhaustive-deps
  [indicatorSize]); //#endregion
  // renders

  const renderIcon = (0, _react.useCallback)(() => {
    if (item.icon) {
      const IconComponent = item.icon;
      return /*#__PURE__*/_react.default.createElement(_reactNativeReanimated.default.View, {
        style: iconStyle
      }, typeof IconComponent === 'function' ? IconComponent({
        size: indicatorSize / 2
      }) : /*#__PURE__*/_react.default.createElement(IconComponent, {
        size: indicatorSize / 2
      }));
    }

    return null;
  }, [item, indicatorSize, iconStyle]);
  return /*#__PURE__*/_react.default.createElement(_reactNativeReanimated.default.View, {
    style: containerStyle
  }, /*#__PURE__*/_react.default.createElement(_reactNativeSvg.Svg, {
    width: indicatorSize,
    height: indicatorSize,
    viewBox: "0 0 ".concat(indicatorSize, " ").concat(indicatorSize)
  }, /*#__PURE__*/_react.default.createElement(AnimatedCircle, {
    r: animatedRadius,
    cx: indicatorSize / 2,
    cy: indicatorSize / 2 // @ts-ignore
    ,
    fill: indicatorBackgroundColor,
    fillOpacity: animatedCircleFillOpacity,
    stroke: indicatorBorderColor,
    strokeWidth: BORDER_WIDTH
  })), renderIcon());
};

const Indicator = /*#__PURE__*/(0, _react.memo)(IndicatorComponent, _lodash.default);
var _default = Indicator;
exports.default = _default;
