// @gorhom/paper-onboarding/lib/commonjs/components/backgroundCircle/BackgroundCircle.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNativeSvg = require("react-native-svg");

var _reactNativeReanimated = _interopRequireWildcard(require("react-native-reanimated"));

var _lodash = _interopRequireDefault(require("lodash.isequal"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

// @ts-ignore 😞
const AnimatedCircle = _reactNativeReanimated.default.createAnimatedComponent(_reactNativeSvg.Circle);

const BackgroundCircleComponent = ({
  index,
  animatedIndex,
  color,
  extendedSize,
  bottomPosition,
  indicatorSize,
  animatedIndicatorsContainerPosition
}) => {
  //#region variables
  //#endregion
  //#region animations
  const animatedFocus = (0, _reactNativeReanimated.interpolate)(animatedIndex, {
    inputRange: [index - 1, index, index + 1],
    outputRange: [0, 1, 2],
    extrapolate: _reactNativeReanimated.Extrapolate.CLAMP
  });
  const animatedRadius = (0, _reactNativeReanimated.interpolate)(animatedFocus, {
    inputRange: [0, 1],
    outputRange: [0, extendedSize],
    extrapolate: _reactNativeReanimated.Extrapolate.CLAMP
  });
  const animatedLeftPosition = (0, _reactNativeReanimated.add)(animatedIndicatorsContainerPosition, indicatorSize / 2, index * indicatorSize); //#endregion
  // render

  return /*#__PURE__*/_react.default.createElement(AnimatedCircle, {
    r: animatedRadius,
    cy: bottomPosition,
    cx: animatedLeftPosition,
    fill: color
  });
};

const BackgroundCircle = /*#__PURE__*/(0, _react.memo)(BackgroundCircleComponent, _lodash.default);
var _default = BackgroundCircle;
exports.default = _default;
