// @gorhom/paper-onboarding/lib/module/components/backgroundCircle/BackgroundCircle.js
import React, { memo } from 'react';
import { Circle } from 'react-native-svg';
import Animated, { interpolate, add, Extrapolate } from 'react-native-reanimated'; // @ts-ignore 😞

import isEqual from 'lodash.isequal';
const AnimatedCircle = Animated.createAnimatedComponent(Circle);

const BackgroundCircleComponent = ({
  index,
  animatedIndex,
  color,
  extendedSize,
  bottomPosition,
  indicatorSize,
  animatedIndicatorsContainerPosition
}) => {
  //#region variables
  //#endregion
  //#region animations
  const animatedFocus = interpolate(animatedIndex, {
    inputRange: [index - 1, index, index + 1],
    outputRange: [0, 1, 2],
    extrapolate: Extrapolate.CLAMP
  });
  const animatedRadius = interpolate(animatedFocus, {
    inputRange: [0, 1],
    outputRange: [0, extendedSize],
    extrapolate: Extrapolate.CLAMP
  });
  const animatedLeftPosition = add(animatedIndicatorsContainerPosition, indicatorSize / 2, index * indicatorSize); //#endregion
  // render

  return /*#__PURE__*/React.createElement(AnimatedCircle, {
    r: animatedRadius,
    cy: bottomPosition,
    cx: animatedLeftPosition,
    fill: color
  });
};

const BackgroundCircle = /*#__PURE__*/memo(BackgroundCircleComponent, isEqual);
export default BackgroundCircle;
